### ISSUE TITLE
Ledger Nano S Hardware Wallet Support

### ISSUE THREAD

**Issue by ryanRfox** [bitshares/bitshares-core#1340]
**User Story**
As a `Ledger Nano S user` I want an implementation to `support BitShares accounts` so that I may `safely store my private keys, sign transactions and submit them to the network`. 

**Additional Context**
- [ ] Hardware Wallet Support (General)
- [ ] DasCoin Implementation

## CORE TEAM TASK LIST
- [ ] Evaluate / Prioritize Feature Request
- [ ] Refine User Stories / Requirements
- [ ] Define Test Cases
- [ ] Design / Develop Solution
- [ ] Perform QA/Testing
- [ ] Update Documentation

**Comment by christophersanborn** [comment]
@marcialvieira and I would like to take a crack at this one.

**Comment by clockworkgr** [comment]
> @marcialvieira and I would like to take a crack at this one. @ryanRfox

fuck yeah!

It shouldn't be TOO hard if you use the EOS one as a starting point...I would've tried it myself but i'm not a C/C++ coder.

**Comment by christophersanborn** [comment]
Adding a task for myself:

- [ ] Do ten hours of immersive background research on the Ledger Nano S libraries and developer docs, to include, if possible compiling and loading the device components of one or both of the two wallets linked above, in order to identify which problems remain to be solved to get a BitShares app running.

**Comment by oxarbitrage** [comment]
@christophersanborn I think that is reasonable.

**Comment by christophersanborn** [comment]
Declaring the first task, "Do ten hours of immersive research" to be complete.  Specific products of this research are:
* Various discussions and information posted to the Telegram group dedicated to the developers on this task.  
* Transcripts of successful interaction between host and Nano S using ledger-app-eos as an example.
* Familiarity with codebase in ledger-app-eos, (a logical starting point for the Bitshares Ledger app).
* Forked repository in [christophersanborn/ledger-app-bitshares](https://github.com/christophersanborn/ledger-app-bitshares) (forked from ledger-app-eos)
* Five initial [issues](https://github.com/christophersanborn/ledger-app-bitshares/issues?q=is%3Aissue+is%3Aopen+sort%3Acreated-asc) in that repository to guide initial development efforts of ledger-app-bitshares.

Tagging: @ryanRfox

**Comment by christophersanborn** [comment]
## Overall Roadmap:

Goal is to have both a Nano S app that runs on the device, allowing any wallet developer to add support, and, at minimum, a simple GUI App to submit to the Ledger App Store to facilitate safe storage and transfer of assets.  (I.e. basic "wallet" functionality).  Beyond that, we may also work on adding support to the cli_wallet and/or providing the [bitshares-ui](https://github.com/bitshares/bitshares-ui) team with the resources that they would need to add support to the reference GUI wallet.

- [ ] Phase I:  Implementation of Nano app and basic developer utilities
  - [ ] Device side: Nano app — @christophersanborn 
  - [ ] Host side: Python scripts — @marcialvieira
- [ ] Phase II:  User interface
  - [ ] Primary Goal: A standalone GUI app for the Ledger app store providing basic wallet functionality
  - [ ] Secondary: Integration into cli_wallet
  - [ ] Secondary: Documentation and maybe some initial interface code to allow bitshares-ui team to integrate into Bitshares GUI 

For the time being, work is being done in [christophersanborn/ledger-app-bitshares](https://github.com/christophersanborn/ledger-app-bitshares), and detailed task management in [christophersanborn/ledger-app-bitshares/issues](https://github.com/christophersanborn/ledger-app-bitshares/issues).

**Comment by clockworkgr** [comment]
> ## Overall Roadmap:
> Goal is to have both a Nano S app that runs on the device, allowing any wallet developer to add support, and, at minimum, a simple GUI App to submit to the Ledger App Store to facilitate safe storage and transfer of assets. (I.e. basic "wallet" functionality). Beyond that, we may also work on adding support to the cli_wallet and/or providing the [bitshares-ui](https://github.com/bitshares/bitshares-ui) team with the resources that they would need to add support to the reference GUI wallet.
> 
> * [ ]  Phase I:  Implementation of Nano app and basic developer utilities
>   
>   * [ ]  Device side: Nano app — @christophersanborn
>   * [ ]  Host side: Python scripts — @marcialvieira
> * [ ]  Phase II:  User interface
>   
>   * [ ]  Primary Goal: A standalone GUI app for the Ledger app store providing basic wallet functionality
>   * [ ]  Secondary: Integration into cli_wallet
>   * [ ]  Secondary: Documentation and maybe some initial interface code to allow bitshares-ui team to integrate into Bitshares GUI
> 
> For the time being, work is being done in [christophersanborn/ledger-app-bitshares](https://github.com/christophersanborn/ledger-app-bitshares), and detailed task management in [christophersanborn/ledger-app-bitshares/issues](https://github.com/christophersanborn/ledger-app-bitshares/issues).

as far as GUI app is concerned, I plan to be adding support for it to Beet once you have the device side working using https://github.com/LedgerHQ/ledgerjs

**Comment by sschiessl-bcp** [comment]
Yes, you are cordially invited to use Beet as your standalone GUI app Christopher :)

**Comment by christophersanborn** [comment]
Nice! @clockworkgr and @sschiessl-bcp; I love the idea of Beet support, and this looks very easy to achieve and opens up so many possibilities.  I've added it to the roadmap above.

**Comment by clockworkgr** [comment]
Nice. To be honest, if you finish the on-device work I should be able to handle the JS side myself. ofcourse, any help would be greatly appreciated :)

**Comment by marcialvieira** [comment]
> >   * [ ]   Host side: Python scripts — @marcialvieira
Estimating between 10 and 20 hours for this component, but may be less because of the similarities.

**Comment by ryanRfox** [comment]
Thanks @christophersanborn and @marcialvieira for your estimates and roadmap. I'v updated the ore Team Task List to reflect them. Please continue with your design and development efforts.

**Comment by dls-cipher** [comment]
As user of Nano Ledger, would love to see BTS supported on it. Keep up the great work and ideas!

**Comment by christophersanborn** [comment]
Logging 10 hours complete towards target "Device Side: Nano app" within the Week 48–49 period.

**Comment by cloud-8** [comment]
Would love to see this come to Trezor also, given it is open hardware aswell as open software.

**Comment by christophersanborn** [comment]
Logging 15 hours toward target "Device Side: Nano app" within the Week 50–51 time period.  (Bringing total thus far toward the target to 25 hours.)  Tagging: @ryanRfox

**Comment by ryanRfox** [comment]
Thanks @christophersanborn for the update.

**Comment by christophersanborn** [comment]
Logging 30 hours toward target "Device Side: Nano app"  in the time period from 1/1/2019 to 1/27/2019.  This covers [21 commits](https://github.com/christophersanborn/ledger-app-bitshares/commits/master) by me from b41063b up through e9f572c.

This brings total thus far toward the target to 55 hours.

Tagging: @ryanRfox

**Comment by marcialvieira** [comment]
Logging 24 hours toward target "Host Side: Nano S implementation" in the time period from 1/18/2019 to 2/4/2019. This covers [8 commits](https://github.com/christophersanborn/ledger-app-bitshares/commits/master?author=marcialvieira) by me from d4e2ece up through 5cfbf3b.

Estimating between 40 and 60 hours for this component, including the 24 hours already done, depending on the need of [operations](https://github.com/bitshares/python-bitshares/blob/master/bitsharesbase/operationids.py), I'm expecting that the complexity decreases a lot by similarities between operations, already implemented operations: transfer, limit_order_create, limit_order_cancel and account_update.

Tagging: @ryanRfox

**Comment by ryanRfox** [comment]
I'm adding this to our Feature Release project board for visual reference. This does **not** have any impact on the release nor is it held the schedule. It's just may way to show the Community "we **are** working on it".
